packages/core/src/common/uri-command-handler.ts

```diff
@@ -47,82 +46,128 @@ export interface MultiUriCommandHandler extends UriCommandHandler<URI[]> {
-        if (args && args[0] instanceof URI) {
-            // @ts-ignore we want to always return URIs
-            return this.isMulti() ? [args[0]] : args[0];
-        }
-        const { selection } = this.selectionService;
-        if (!this.isMulti()) {
-            return UriSelection.getUri(selection) as T;
-        }
-        const uris = UriSelection.getUris(selection);
-        return uris as T;
-    }
+    protected abstract getArgsWithURI(...args: any[]): [T | undefined, ...any[]];
 
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
     execute(...args: any[]): object | undefined {
-        const uri = this.getUri(...args);
-        return uri ? this.handler.execute(uri, ...args) : undefined;
+        const [uri, ...others] = this.getArgsWithURI(...args);
+        return uri ? this.handler.execute(uri, ...others) : undefined;
     }
 
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
     isVisible(...args: any[]): boolean {
-        const uri = this.getUri(...args);
+        const [uri, ...others] = this.getArgsWithURI(...args);
         if (uri) {
             if (this.handler.isVisible) {
-                return this.handler.isVisible(uri as T, ...args);
+                return this.handler.isVisible(uri, ...others);
             }
             return true;
         }
         return false;
     }
 
-    // eslint-disable-next-line @typescript-eslint/no-explicit-any
     isEnabled(...args: any[]): boolean {
-        const uri = this.getUri(...args);
+        const [uri, ...others] = this.getArgsWithURI(...args);
         if (uri) {
             if (this.handler.isEnabled) {
-                return this.handler.isEnabled(uri as T, ...args);
+                return this.handler.isEnabled(uri, ...others);
             }
             return true;
         }
         return false;
     }
+}
+
+export class MonoSelectUriCommandHandler extends AbstractUriAwareCommandHandler<URI> {
+    protected getArgsWithURI(...args: any[]): [URI | undefined, ...any[]] {
+        const [maybeURI, ...others] = args;
+        if (maybeURI instanceof URI) {
+            return [maybeURI, ...others];
+        }
+        const { selection } = this.selectionService;
+        const URIFromSelection = UriSelection.getUri(selection);
+        return [URIFromSelection, ...args];
+    }
+}
+
+export class MultiSelectUriCommandHandler extends AbstractUriAwareCommandHandler<URI[]> {
+    protected getArgsWithURI(...args: any[]): [URI[] | undefined, ...any[]] {
+        const [maybeURIs, ...others] = args;
+        if (Array.isArray(maybeURIs) && maybeURIs.every(uri => uri instanceof URI)) {
+            return [maybeURIs, ...others];
+        }
+        const { selection } = this.selectionService;
+        const URIsFromSelection = UriSelection.getUris(selection);
+        return [URIsFromSelection, ...args];
+    }
+}
+
+export class UriAwareCommandHandler<T extends MaybeArray<URI>> extends AbstractUriAwareCommandHandler<T> {
```

Comment:
I tried playing with conditional typing, but idk what to think: [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgSgSSgXigewEYCsIGNgDcAUKJFALICGIGEAggE4PUA8AKgHwpRtQA+PANoBdIkQgAPMGgbAouADaUAzstgMAlnQDulBhADCaALbHKAOwAmACQuWFEBuyiTgEK2qo16TVvAQcXADeRFBh8mjmysAMAK74MgAUAPpoYMAakcoAXOpauvpGpna2Vg4MAHQA8umZUewcAJRQIeFtUAD0HVAAtD3K5hpgfaHhAL5EE5LSslDmlMYQymCUuNAwmjp6hiZmVqX2ji2jYdMychrmbgwAZqvQNRlZzq7ulp7UtIzMICz+gcd2uFjLEFBlcrxXh5YIgRFAAPxQGKxaC5O4KZQQE5QCYTIhddAAayI5gg2jyW0KuxKdnKf0QHESQSgILBGjRlAx0DGjWIpPJG3y2yKexstMc9IQIkZzNZ4KRcW5vLEBIAhGhifyKQUdsV9uKnP8mSzQfLkUq+WTtcLqfqyhL-NLjXL2VB0ZiccqgA)

Maybe this is actually a low cost fix for the typings of `UriAwareCommandHandler`?

Reply:
Nice! I messed around trying to do something like this, but couldn't get TS to cooperate. To my mind, the main argument in favor of splitting it into two classes is to resolve the possible contradiction between the generic and the options, and if we can do that with conditional typing, I don't see a lot to recommend the two-class solution. What do you think?

Reply:
I would have avoided complexifying the typings if possible, but since it fixes the old API I am fine with it!

Reply:
I guess it has the disadvantage that if anyone has declared something like
```
const myHandlerOptions: UriAwareCommandHandler.Options = {multi: true};
```
we'll have to declare a default for the generic of options to avoid breaking their code by requiring a generic where none was needed before, but the default may not agree with what they've specified, so there's no way to _guarantee_ it won't break.

Given that, maybe you're right that it's better not to revamp the typings and just to provide options that don't run into the same complication.

Reply:
It sure sounds like less trouble :)